lib/Dialect/Lattigo/Transforms/AllocToInplace.cpp

```diff
@@ -3,98 +3,225 @@
+  //
+  // However, this might not be optimal in terms of cache-friendliness for
+  // various accelerators. One basic optimization is to use the dead value that
+  // is closest to the current operation in the block. But as we do not have the
+  // information of the memory layout, we do not implement this optimization.
+  Value getAvailableStorage(Operation *op, Liveness *liveness) const {
+    Value availableStorage;
+    for (auto &[storage, values] : storageToReferringValues) {
+      // storage and all referring values are dead
+      if (std::all_of(
+              values.begin(), values.end(),
+              [&](Value value) { return liveness->isDeadAfter(value, op); }) &&
+          liveness->isDeadAfter(storage, op)) {
+        availableStorage = storage;
+        break;
+      }
+    }
+    return availableStorage;
+  }
+
+  void replaceAllocWithInplace(Operation *oldOp, Operation *newOp,
+                               Value storage) {
+    // add newly created referring value
+    for (auto result : newOp->getResults()) {
+      addReferringValue(storage, result);
+    }
+    // remove storage of old op
+    for (auto result : oldOp->getResults()) {
+      mergeStorage(result, storage);
+    }
+  }
+
+ private:
+  DenseMap<Value, SmallVector<Value>> storageToReferringValues;
+};
+
 template <typename BinOp, typename InplaceOp>
 struct ConvertBinOp : public OpRewritePattern<BinOp> {
   using OpRewritePattern<BinOp>::OpRewritePattern;
 
-  ConvertBinOp(mlir::MLIRContext *context, Liveness *liveness)
-      : OpRewritePattern<BinOp>(context), liveness(liveness) {}
+  ConvertBinOp(mlir::MLIRContext *context, Liveness *liveness,
+               DenseMap<Block *, StorageInfo> *blockToStorageInfo)
+      : OpRewritePattern<BinOp>(context),
+        liveness(liveness),
+        blockToStorageInfo(blockToStorageInfo) {}
 
   LogicalResult matchAndRewrite(BinOp op,
                                 PatternRewriter &rewriter) const override {
-    // operand 0 is evaluator
-    auto lhs = op.getOperand(1);
-    if (!liveness->isDeadAfter(lhs, op)) {
+    auto &storageInfo = (*blockToStorageInfo)[op->getBlock()];
+    auto storage = storageInfo.getAvailableStorage(op, liveness);
+    if (!storage) {
       return failure();
     }
 
-    // InplaceOp has the form: output = InplaceOp(evaluator, lhs, rhs, inplace)
-    // where inplace is the actual output but for SSA form we need to return a
-    // new value
-    rewriter.replaceOpWithNewOp<InplaceOp>(op, op.getOperand(1).getType(),
-                                           op.getOperand(0), op.getOperand(1),
-                                           op.getOperand(2), op.getOperand(1));
+    // InplaceOp has the form: output = InplaceOp(evaluator, lhs, rhs,
+    // inplace) where inplace is the actual output but for SSA form we need to
+    // return a new value
+    auto inplaceOp = rewriter.replaceOpWithNewOp<InplaceOp>(
```

Comment:
I patched the problem internally, so no changes are needed on this PR, but here was the problem: you were replacing the op before calling `replaceAllocWithInplace`, which then queries the (now deleted) `oldOp`'s results.

For whatever reason, externally that deletion was not yet overwritten so it works, but internally the memory was cleared and it led to `oldOp->getNumResults()` being like 28376422345, and hitting a timeout while looping over the "results" (or hitting a memory error when trying to access the iterator used for `reserve`).

So this PR will land with a patch that creates the op, then calls `replaceAllocWithInplace`, then does the replacement (in each of the three patterns).